pkg/curatedpackages/packagecontrollerclient.go

```diff
@@ -49,19 +60,102 @@ type PackageControllerClient struct {
 	// activeBundleTimeout is the timeout to activate a bundle on installation.
 	activeBundleTimeout time.Duration
 	valuesFileWriter    filewriter.FileWriter
+	// skipWaitForPackageBundle indicates whether the installer should wait
+	// until a package bundle is activated.
+	//
+	// Skipping the wait is desirable for full cluster lifecycle use cases,
+	// where resource creation and error reporting are asynchronous in nature.
+	skipWaitForPackageBundle bool
+	// tracker creates k8s clients for workload clusters managed via full
+	// cluster lifecycle API.
+	clientBuilder ClientBuilder
+
+	// mu provides some thread-safety.
+	mu sync.Mutex
+}
+
+// ClientBuilder returns a k8s client for the specified cluster.
+type ClientBuilder interface {
+	GetClient(context.Context, types.NamespacedName) (client.Client, error)
 }
 
 type ChartInstaller interface {
-	InstallChart(ctx context.Context, chart, ociURI, version, kubeconfigFilePath, namespace, valueFilePath string, values []string) error
+	InstallChart(ctx context.Context, chart, ociURI, version, kubeconfigFilePath, namespace, valueFilePath string, skipCRDs bool, values []string) error
+}
+
+// ChartInstallationDeleter handles deleting chart installations.
+type ChartInstallationDeleter interface {
+	Delete(ctx context.Context, kubeconfigFilePath, installName, namespace string) error
+}
+
+// NewPackageControllerClientFullLifecycle creates a PackageControllerClient
+// for the Full Cluster Lifecycle controller.
+//
+// It differs because the CLI use case has far more information available at
+// instantiation, while the FCL use case has less information at
+// instantiation, and the rest when cluster creation is triggered.
+func NewPackageControllerClientFullLifecycle(logger logr.Logger, chartInstaller ChartInstaller, deleter ChartInstallationDeleter, kubectl KubectlRunner, clientBuilder ClientBuilder) *PackageControllerClient {
+	return &PackageControllerClient{
+		chartInstaller:           chartInstaller,
+		deleter:                  deleter,
+		kubectl:                  kubectl,
+		skipWaitForPackageBundle: true,
+		eksaRegion:               eksaDefaultRegion,
+		clientBuilder:            clientBuilder,
+	}
+}
+
+// EnableFullLifecycle wraps Enable to handle run-time arguments.
+//
+// This method fills in the gaps between the original CLI use case, where all
+// information is known at PackageControllerClient initialization, and the
+// Full Cluster Lifecycle use case, where there's limited information at
```

Comment:
What are some examples of this limited information?

Reply:
Basically, if you look at `NewPackageControllerClient`'s parameters, and subtract these parameters out, you're left with the set of parameters that full cluster lifecycle doesn't know at instantiation. Things like the cluster name. Some of this is because of different timings, and some is because when used by full cluster lifecycle, this package controller client might be reused to build multiple clusters, whereas in a CLI usage, it builds a single cluster, then exits.

There's probably a lot of refactoring that could be done here, but this PR was big enough as is.

Reply:
I'll add something to the comment.